test/e2e/performance_test.go

```diff
@@ -0,0 +1,349 @@
+	err = waitNetworkPolicyRealize(workloadsNum, data)
+	if err != nil {
+		b.Fatalf("Checking network policies realization failed: %v", err)
+	}
+	b.Log("Network policy realized")
+
+	serverURL := &url.URL{Scheme: "http", Host: nginxPodIP, Path: "/"}
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		b.Logf("Running http request bench %d/%d", i+1, b.N)
+		cmd := []string{"ab", "-n", fmt.Sprint(times), "-c", fmt.Sprint(*httpConcurrency), serverURL.String()}
+		stdout, stderr, err := data.runCommandFromPod(testNamespace, benchABPodName, abContainerName, cmd)
+		if err != nil {
+			b.Errorf("Error when running http request %dx: %v, stdout: %s, stderr: %s\n", times, err, stdout, stderr)
+		}
+	}
+}
+
+// networkPolicyRealize runs the benchmark of the time cost of a network policy with `workloadsNum` amount CIDRs
+// to be realized into flow entries. In order to have entities for the network policy to apply to, we
+// create two dummy Pods: apache-bench and Nginx, they don't have activity during the benchmark test.
+func networkPolicyRealize(workloadsNum int, data *TestData, b *testing.B) {
+	setupTestPods(data, b)
+	for i := 0; i < b.N; i++ {
+		go func() {
+			err := populateWorkloads(generateWorkloads(workloadsNum), data)
+			if err != nil {
+				b.Fatalf("Error when populating workload: %v", err)
+			}
+		}()
+
+		b.Log("Waiting the network policy to be realized")
+		b.StartTimer()
+		err := waitNetworkPolicyRealize(workloadsNum, data)
+		if err != nil {
+			b.Fatalf("Checking network policies realization failed: %v", err)
+		}
+		b.StopTimer()
+		b.Log("Network policy realized")
+
+		err = data.clientset.NetworkingV1().NetworkPolicies(testNamespace).Delete(workloadNetworkPolicyName, new(metav1.DeleteOptions))
+		if err != nil {
+			b.Fatalf("Error when cleaning up network policies after running one bench iteration: %v", err)
+		}
+	}
+}
+
+func dumpFlows(data *TestData) (string, error) {
+	antreaPodName, err := data.getAntreaPodOnNode(masterNodeName())
+	if err != nil {
+		return "", err
+	}
+	cmd := []string{"ovs-ofctl", "dump-flows", "br-int"}
+	stdout, _, err := data.runCommandFromPod(AntreaNamespace, antreaPodName, "antrea-agent", cmd)
+	if err != nil {
+		return "", err
+	}
+	return stdout, nil
+}
+
+func waitNetworkPolicyRealize(workloads int, data *TestData) error {
+	return wait.PollImmediate(0, *realizeTimeout, func() (bool, error) {
+		return checkRealize(workloads, data)
+	})
+}
+
+// checkRealize checks if all CIDRs in the workload network policy are realized.
+// Since `countFlows` is an SSH operation, we could not get the precise time of the realization. According to the
+// antrea-agent implementation, each CIDR in a network policy will be reflected in one flow entry. To reduce the
+// over counted duration which is introduced by the checking, this function ignores a constant amount of flow
+// entries that are installed by antrea-agent.
+func checkRealize(workloadsNum int, data *TestData) (bool, error) {
```

Comment:
suggestion for comment:
> checkRealize checks if all CIDR rules in the Network Policy have been realized as OVS flows. It calls countFlows, which counts the number of flows installed in the OVS bridge of the master Node. Since it is done over SSH, the time measurement is not completely accurate. checkRealize returns true when the number of flows exceeds the number of CIDR rules. This relies on the implementation knowledge that given a single policy, the Antrea agent will install exactly one flow per CIDR rule. We ignore the non-policy related flows installed by Antrea (small number), which yet again may introduce some inaccuracy.

Reply:
BTW, maybe we should only dump flows in the relevant table (is that the IngressRuleTable?). This way we don't have to worry about other flows.

Reply:
Yes, IngressRuleTable, this would be better. I changed accordingly, the comment is applied too.